src/db_reader.rs

```diff
@@ -0,0 +1,816 @@
+use crate::iter::KeyValueIterator;
+use crate::manifest::Manifest;
+use crate::manifest_store::{ManifestStore, StoredManifest};
+use crate::mem_table::{ImmutableMemtable, VecDequeKeyValueIterator, WritableKVTable};
+use crate::reader::Reader;
+use crate::sorted_run_iterator::SortedRunIterator;
+use crate::sst::SsTableFormat;
+use crate::sst_iter::{SstIterator, SstIteratorOptions};
+use crate::tablestore::TableStore;
+use crate::{filter, utils, wal_replay, Checkpoint, DbIterator};
+use bytes::Bytes;
+use log::{info, warn};
+use object_store::path::Path;
+use object_store::ObjectStore;
+use parking_lot::{Mutex, RwLock};
+use serde::{Deserialize, Serialize};
+use std::collections::VecDeque;
+use std::ops::{Add, RangeBounds, Sub};
+use std::sync::Arc;
+use std::time::{Duration, SystemTime};
+use std::{cmp, mem};
+use tokio::runtime::Handle;
+use tokio::select;
+use tokio::sync::mpsc::{UnboundedReceiver, UnboundedSender};
+use uuid::Uuid;
+
+/// Read-only interface for accessing a database from either
+/// the latest persistent state or from an arbitrary checkpoint.
+pub struct DbReader {
+    inner: Arc<DbReaderInner>,
+    manifest_poller: Option<ManifestPoller>,
+}
+
+struct DbReaderInner {
+    manifest_store: Arc<ManifestStore>,
+    table_store: Arc<TableStore>,
+    options: DbReaderOptions,
+    state: Arc<RwLock<CheckpointState>>,
+}
+
+struct ManifestPoller {
+    join_handle: Mutex<Option<tokio::task::JoinHandle<Result<(), SlateDBError>>>>,
+    thread_tx: UnboundedSender<ManifestPollerMsg>,
+}
+
+enum ManifestPollerMsg {
+    Shutdown,
+}
+
+#[derive(Clone)]
+pub(crate) struct CheckpointState {
+    pub(crate) checkpoint: Checkpoint,
+    pub(crate) manifest: Manifest,
+    pub(crate) imm_memtables: Vec<Arc<ImmutableMemtable>>,
+}
+
+#[derive(Clone, Deserialize, Serialize)]
+pub struct DbReaderOptions {
+    /// How frequently to poll for new manifest files. Refreshing the manifest file allows readers
+    /// to detect newly compacted data.
+    pub manifest_poll_interval: Duration,
+
+    /// For readers that refresh their checkpoint, this specifies the lifetime to use for the
+    /// created checkpoint. The checkpoint’s expire time will be set to the current time plus
+    /// this value. If not specified, then the checkpoint will be created with no expiry, and
+    /// must be manually removed. This lifetime must always be greater than
+    /// manifest_poll_interval x 2
+    pub checkpoint_lifetime: Option<Duration>,
+
+    /// The max size of a single in-memory table used to buffer WAL entries
+    /// Defaults to 64MB
+    pub max_memtable_bytes: u64,
+
+    #[serde(skip)]
+    pub block_cache: Option<Arc<dyn DbCache>>,
+
+    /// The Clock to use for manifest polling and checkpoint lifetime maintenance.
+    ///
+    /// Default: the default clock uses the local system time on the machine
+    #[serde(skip)]
+    #[serde(default = "crate::config::default_clock")]
+    pub clock: Arc<dyn Clock + Send + Sync>,
```

Comment:
Something feels off about this. For the manifest polling and checkpoint lifetimes, we always want to use system (wall-clock) time. But I think we're [now filtering expired records](https://github.com/slatedb/slatedb/pull/458) using a user-supplied clock. I think the `Clock` stuff you're doing should always use SystemTime or just the std clock. OTOH, I think the reads should use a user-supplied clock. Note that in the PR I linked to, has a `get_now_for_read` that's fed into `filter`, which is fed into `unwrap_result`.

Reply:
Yeah, done a bit hastily. I was looking for a good way to control the clock in tests. I'll find another way.